contracts/sales/FixedPriceSale.sol

```diff
@@ -205,69 +179,44 @@ contract FixedPriceSale {
     }
 
     /// @dev realease tokenIn back to investors if minimumRaise not reached
-    function releaseTokens() public {
+    /// can also be used from external script to automatically release tokens for investors
+    function releaseTokens(address account) public {
         require(minimumRaise > 0, "FixedPriceSale: no minumumRaise");
         require(
             block.timestamp > endDate,
             "FixedPriceSale: endDate not passed"
         );
         require(
-            tokensPurchased[msg.sender] > 0,
+            tokensPurchased[account] > 0,
             "FixedPriceSale: no tokens purchased by this investor"
         );
         require(
             tokensSold < minimumRaise,
             "FixedPriceSale: minumumRaise reached"
         );
 
-        uint256 tokensAmount = tokensPurchased[msg.sender];
-        tokensPurchased[msg.sender] = 0;
-        isClosed = true;
-        TransferHelper.safeTransfer(address(tokenIn), msg.sender, tokensAmount);
-        emit NewTokenRelease(msg.sender, tokensAmount);
+        uint256 tokensAmount = tokensPurchased[account];
+        tokensPurchased[account] = 0;
+        TransferHelper.safeTransfer(address(tokenIn), account, tokensAmount);
+        emit NewTokenRelease(account, tokensAmount);
     }
 
     /// @dev let investors claim their purchased tokens
-    function claimTokens() public {
+    /// can also be used from external script to automatically claim tokens for investors
+    function claimTokens(address account) public {
         require(isClosed, "FixedPriceSale: sale not closed");
         require(
-            tokensPurchased[msg.sender] > 0,
+            tokensPurchased[account] > 0,
             "FixedPriceSale: no tokens to claim"
         );
-        uint256 purchasedTokens = tokensPurchased[msg.sender];
-        tokensPurchased[msg.sender] = 0;
+        uint256 purchasedTokens = tokensPurchased[account].mul(tokenPrice);
```

Comment:
I am pretty sure this calculation should be in `releaseTokens()` since the `tokensPurchased` mapping is denominated in `tokensOut` not `tokensIn`.  `claimTokens()` should transfer the amount directly from `tokensPurchased`.

At least this is how the current SC and UI treat it.
The amount parameter in `buyTokens()` is what gets saved to the mapping which is `tokensOut` on the UI side and directly compared to `allocationMin` so must be `tokensOut`.

We should consider renaming this mapping to something clearer maybe.

Reply:
on buyTokens()

tokenIn token are send to the contract and added to the tokensPurchased. Think dai. Then dai * price gives you the token out amount. As example  1 dai tokenprice is 2 $kn per dai . So 1 * 2 = 2 $tkn

![Screenshot at 2021-06-22 16-43-40](https://user-images.githubusercontent.com/918180/122945752-144c9980-d379-11eb-97b9-3df57434e787.png)

tokensOut on the UI side means in my example we send how much $tkn we want, not how much we invest? 

We have to change this on the UI side I guess. because we calculate this in the frontend and this may not match the sent tokenIn.

Reply:
On the UI side and current SC we are always sending tokensOut to buyTokens().
If we do it the other way around it would need constantly converted and changes basically everywhere, in UI and SC.
Since amounts like minRaise and allocation amounts are in tokensOut it is best to keep the calculations to a minimum so tokensOut makes more sense.